libtasn1/doc/asn1Decoding.1
.\" DO NOT MODIFY THIS FILE!  It was generated by help2man 1.47.3.
.TH ASN1DECODING "1" "July 2016" "asn1Decoding (libtasn1) 4.9" "User Commands"
.SH NAME
asn1Decoding \- ASN.1 DER decoder
.SH SYNOPSIS
.B asn1Decoding
[\fI\,OPTION\/\fR] \fI\,DEFINITIONS ENCODED ASN1TYPE\/\fR
.SH DESCRIPTION
Decodes DER data in ENCODED file, for the ASN1TYPE element
described in ASN.1 DEFINITIONS file, and print decoded structures.
.TP
\fB\-b\fR, \fB\-\-benchmark\fR
perform a benchmark on decoding
.TP
\fB\-s\fR, \fB\-\-strict\fR
use strict DER decoding
.TP
\fB\-h\fR, \fB\-\-help\fR
display this help and exit
.TP
\fB\-v\fR, \fB\-\-version\fR
output version information and exit
.SH AUTHOR
Written by Fabio Fiorina.
.SH "REPORTING BUGS"
Report bugs to: help\-libtasn1@gnu.org
.br
GNU Libtasn1 home page: <http://www.gnu.org/software/libtasn1/>
.br
General help using GNU software: <http://www.gnu.org/gethelp/>
.SH COPYRIGHT
Copyright \(co 2014 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>.
.br
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.
.SH "SEE ALSO"
The full documentation for
.B asn1Decoding
is maintained as a Texinfo manual.  If the
.B info
and
.B asn1Decoding
programs are properly installed at your site, the command
.IP
.B info libtasn1
.PP
should give you access to the complete manual.


NEWS
Changes between 0.3.0 and 0.3.1:
--------------------------------

  * Fix crash when used from multipe threads
  * Fix nl_langinfo detection on Android
  * Fix build for winstore


Changes between 0.2.7 and 0.3.0:
--------------------------------

  * Implement SMB SESSION LOGOFF (0x74)
  * Add smb_stat_list_next() to reduce directory traversal speed
  * Fix guest login
  * Fix compilation when using compatibility layer
  * Fix crash against malformed packet in smb_fstat
  * Fix too small buffer in smb_session_login_spnego, leading to failure
  * Fix payload for smb_negotiate for arm5te
  * Fix misc crashes


Changes between 0.2.6 and 0.2.7:
--------------------------------

  * Fix build when lacking getifaddrs


Changes between 0.2.5 and 0.2.6:
--------------------------------

  * Fix connection with a smb server listening on the port 139
  * Fix compilation and struct packing for MSVC
  * Fix sample code
  * Fix Windows broadcast addresses on multiple interfaces
  * Fix return status on smb_fstat() call


Changes between 0.2.4 and 0.2.5:
--------------------------------

 * Add cancellation support without pipe()
 * Continued port to Windows
 * Fix teardown crash and wrong asserts
 * Fix potential crash
 * Minor build system cleanup


Changes between 0.2.3 and 0.2.4:
--------------------------------

 * Fix compilation for Android
 * Continued port to Windows


Changes between 0.2.2 and 0.2.3:
--------------------------------

 * Partial port to Windows (Win32 and WinRT/UWP)
 * Fix logged state after smb_session_login_ntlm


Changes between 0.2.1 and 0.2.2:
--------------------------------

 * Disable DEBUG mode by default
 * Stability improvements


Changes between 0.2.0 and 0.2.1:
--------------------------------

 * Fix seek with file > 2GB on Android
 * smb_fread can skip bytes


Changes between 0.1.0 and 0.2.0:
--------------------------------

 * Add files and folder deletion with smb_directory_rm and smb_file_rm
 * Add folder creation with smb_directory_create
 * Add files creation with smb_fwrite
 * Add files and folder move/rename with smb_file_mv
 * Fix session state after spnego authentication
 * Fix return values of all public functions: return a negative number in case
   of error, and 0 or a positive number in case of success.
 * Add smb_session_get_nt_status: get the last NT_STATUS
 * Only export needed symbols


Changes between 0.0.7 and 0.1.0:
--------------------------------

 * Fix shares listing when server returns a tid equal to 0
   Notably TimeCapsule and AirPort Express/Extreme servers
 * Fix crashes and wrong assertions
 * Improve BSD compatibility


Changes between 0.0.6 and 0.0.7:
--------------------------------

 * Rework folders opening with large number of files (FIND_NEXT)
 * Fix for reading from OS X SMB Sharing
 * Fix build for old glibc versions
 * Fix support for large files (fix for >1GiB and for >4GiB)
 * Fix potential deadlock when server disconnects


Changes between 0.0.5 and 0.0.6:
--------------------------------

 * Fix iOS compilation
 * Fix regression on smb_buffer_alloca that introduced login and discovery
   issues
 * Fix double-free crash


Changes between 0.0.4 and 0.0.5:
--------------------------------

 * Rework netbios_ns timeouts
 * Expose less APIs and defines
 * Fix support for Darwin OS missing clock_gettime()


Changes between 0.0.3 and 0.0.4:
--------------------------------

 * Improve discovery that is now done in a separate thread and lot more robust
 * API: Add netbios_ns_discover_start and netbios_ns_discover_stop to start and
   stop a discovery task
 * API: Remove netbios_ns_discover, netbios_ns_entry_count, netbios_ns_entry_at,
   netbios_ns_clear, netbios_ns_abort. Use the new discovers calls instead
 * Fix smb_fopen failing to open long path


Changes between 0.0.2 and 0.0.3:
--------------------------------

 * Port to Android
 * Simplification of the NetBios discovery API
 * New abort mechanism when discovery: netbios_ns_abort
 * Numerous cleanups, include headers and API
 * Provide a replacement for strlcpy
 * Fix numerous warnings, build issues and leaks
 * Support SMBX servers


Changes between 0.0.1 and 0.0.2:
--------------------------------

 * Fix various build failures


First release 0.0.1:
--------------------

 * This is the first basic release.
 * It should provide the basics for this library, but is not ready yet for
   prime-time


libtasn1/src/asn1Decoding.c
/* asn1Decoding.c ---  program to generate an ASN1 type from a DER coding.
 * Copyright (C) 2002-2014 Free Software Foundation, Inc.
 *
 * This file is part of LIBTASN1.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 *
 */

#include <config.h>

#include <stdio.h>
#include <string.h>
#include <stdlib.h>
#include <unistd.h>
#include <getopt.h>

#include <libtasn1.h>

#include <progname.h>
#include <version-etc.h>
#include <read-file.h>
#include "benchmark.h"

static int decode (asn1_node definitions, const char *typeName, void *der,
		   int der_len, int benchmark, int strict);

/* This feature is available in gcc versions 2.5 and later.  */
#if __GNUC__ < 2 || (__GNUC__ == 2 && __GNUC_MINOR__ < 5)
#define ATTR_NO_RETRUN
#else
#define ATTR_NO_RETRUN __attribute__ ((__noreturn__))
#endif

ATTR_NO_RETRUN static void
usage (int status)
{
  if (status != EXIT_SUCCESS)
    fprintf (stderr, "Try `%s --help' for more information.\n", program_name);
  else
    {
      printf ("\
Usage: %s [OPTION] DEFINITIONS ENCODED ASN1TYPE\n", program_name);
      printf ("\
Decodes DER data in ENCODED file, for the ASN1TYPE element\n\
described in ASN.1 DEFINITIONS file, and print decoded structures.\n\
\n");
      printf ("\
  -b, --benchmark       perform a benchmark on decoding\n\
  -s, --strict          use strict DER decoding\n\
  -h, --help            display this help and exit\n\
  -v, --version         output version information and exit\n");
      emit_bug_reporting_address ();
    }
  exit (status);
}

int
main (int argc, char *argv[])
{
  static const struct option long_options[] = {
    {"help", no_argument, 0, 'h'},
    {"strict", no_argument, 0, 's'},
    {"debug", no_argument, 0, 'd'},
    {"benchmark", no_argument, 0, 'b'},
    {"version", no_argument, 0, 'v'},
    {0, 0, 0, 0}
  };
  int option_index = 0;
  int option_result;
  char *inputFileAsnName = NULL;
  char *inputFileDerName = NULL;
  char *typeName = NULL;
  asn1_node definitions = NULL;
  char errorDescription[ASN1_MAX_ERROR_DESCRIPTION_SIZE];
  int asn1_result = ASN1_SUCCESS;
  unsigned char *der;
  int der_len = 0, benchmark = 0;
  int strict = 0, debug = 0;
  /* FILE *outputFile; */

  set_program_name (argv[0]);

  opterr = 0;			/* disable error messages from getopt */

  while (1)
    {

      option_result =
	getopt_long (argc, argv, "hbdsvc", long_options, &option_index);

      if (option_result == -1)
	break;

      switch (option_result)
	{
	case 'h':		/* HELP */
	  usage (EXIT_SUCCESS);
	  break;
	case 'b':
	  benchmark = 1;
	  break;
	case 'd':
	  debug = 1;
	  break;
	case 's':
	  strict = 1;
	  break;
	case 'v':		/* VERSION */
	  version_etc (stdout, program_name, PACKAGE, VERSION,
		       "Fabio Fiorina", NULL);
	  exit (0);
	  break;
	case '?':		/* UNKNOW OPTION */
	  fprintf (stderr,
		   "asn1Decoding: option '%s' not recognized or without argument.\n\n",
		   argv[optind - 1]);
	  usage (EXIT_FAILURE);
	  break;
	default:
	  fprintf (stderr,
		   "asn1Decoding: ?? getopt returned character code Ox%x ??\n",
		   option_result);
	}
    }

  if (optind == argc || optind == argc - 1 || optind == argc - 2)
    {
      fprintf (stderr, "asn1Decoding: input files or ASN.1 type "
	       "name missing\n");
      usage (EXIT_FAILURE);
    }

  inputFileAsnName = strdup(argv[optind]);
  inputFileDerName = strdup(argv[optind + 1]);
  typeName = strdup(argv[optind + 2]);

  if (!(inputFileAsnName && inputFileDerName && typeName))
    {
      fprintf(stderr, "allocation failed\n");
      free(inputFileAsnName);
      free(inputFileDerName);
      free(typeName);
      exit(1);
    }

  asn1_result =
    asn1_parser2tree (inputFileAsnName, &definitions, errorDescription);

  switch (asn1_result)
    {
    case ASN1_SUCCESS:
      fprintf (stderr, "Parse: done.\n");
      break;
    case ASN1_FILE_NOT_FOUND:
      fprintf (stderr, "asn1Decoding: FILE %s NOT FOUND\n", inputFileAsnName);
      break;
    case ASN1_SYNTAX_ERROR:
    case ASN1_IDENTIFIER_NOT_FOUND:
    case ASN1_NAME_TOO_LONG:
      fprintf (stderr, "asn1Decoding: %s\n", errorDescription);
      break;
    default:
      fprintf (stderr, "libtasn1 ERROR: %s\n", asn1_strerror (asn1_result));
    }

  if (asn1_result != ASN1_SUCCESS)
    {
      free (inputFileAsnName);
      free (inputFileDerName);
      free (typeName);
      exit (1);
    }


  {
    size_t tmplen;
    der = (unsigned char *) read_binary_file (inputFileDerName, &tmplen);
    der_len = tmplen;
  }

  /* read_binary_file() returns a buffer with more data than required,
   * with this reallocation we ensure that memory accesses outside the
   * boundaries are detected */
  if (der != NULL && debug != 0)
    der = realloc(der, der_len);

  if (der == NULL)
    {
      fprintf (stderr, "asn1Decoding: could not read '%s'\n",
	       inputFileDerName);
      asn1_delete_structure (&definitions);

      free (inputFileAsnName);
      free (inputFileDerName);
      free (typeName);
      exit (1);
    }

 /*****************************************/
  /* ONLY FOR TEST                         */
 /*****************************************/
  /*
     der_len=0;
     outputFile=fopen("data.p12","w");
     while(fscanf(inputFile,"%c",der+der_len) != EOF){
     if((der_len>=0x11) && (der_len<=(0xe70)))
     fprintf(outputFile,"%c",der[der_len]);
     der_len++;
     }
     fclose(outputFile);
     fclose(inputFile);
   */

  if (decode (definitions, typeName, der, der_len, benchmark, strict) != ASN1_SUCCESS)
    {
      asn1_delete_structure (&definitions);
      free (inputFileAsnName);
      free (inputFileDerName);
      free (typeName);
      free (der);
      exit (1);
    }

  asn1_delete_structure (&definitions);

  free (der);

  free (inputFileAsnName);
  free (inputFileDerName);
  free (typeName);

  if (asn1_result != ASN1_SUCCESS)
    exit (1);

  exit (0);
}

static int
simple_decode (asn1_node definitions, const char *typeName, void *der,
	       int der_len, int benchmark, int strict)
{

  int asn1_result;
  asn1_node structure = NULL;
  char errorDescription[ASN1_MAX_ERROR_DESCRIPTION_SIZE];

  asn1_result = asn1_create_element (definitions, typeName, &structure);

  /* asn1_print_structure(stdout,structure,"",ASN1_PRINT_ALL); */


  if (asn1_result != ASN1_SUCCESS)
    {
      fprintf (stderr, "Structure creation: %s\n",
	       asn1_strerror (asn1_result));
      asn1_delete_structure (&structure);
      return asn1_result;
    }

  if (strict != 0)
    asn1_result =
      asn1_der_decoding2(&structure, der, &der_len, ASN1_DECODE_FLAG_STRICT_DER, errorDescription);
  else
    asn1_result =
      asn1_der_decoding (&structure, der, der_len, errorDescription);

  if (!benchmark)
    fprintf (stderr, "\nDecoding: %s\n", asn1_strerror (asn1_result));
  if (asn1_result != ASN1_SUCCESS)
    {
      fprintf (stderr, "asn1Decoding: %s\n", errorDescription);
      asn1_delete_structure (&structure);
      return asn1_result;
    }

  if (!benchmark)
    {
      fprintf (stderr, "\nDECODING RESULT:\n");
      asn1_print_structure (stdout, structure, "",
			    ASN1_PRINT_NAME_TYPE_VALUE);
    }
  asn1_delete_structure (&structure);
  return ASN1_SUCCESS;
}

static int
decode (asn1_node definitions, const char *typeName, void *der, int der_len,
	int benchmark, int strict)
{
  struct benchmark_st st;

  if (benchmark == 0)
    return simple_decode (definitions, typeName, der, der_len, benchmark, strict);
  else
    {
      start_benchmark (&st);

      do
	{
	  simple_decode (definitions, typeName, der, der_len, benchmark, strict);
	  st.size++;
	}
      while (benchmark_must_finish == 0);

      stop_benchmark (&st, "structures");
      fprintf (stdout, "\n");

    }
  return ASN1_SUCCESS;
}
